`master` promotion

mrjackv opened the pull request:
Sister PRs:
* [revng-orchestra](https://github.com/revng/revng-orchestra/pull/65)
* [hitman](https://rev.ng/gitlab/revng-private/hitman/-/merge_requests/52)
* [revng](https://github.com/revng/revng/pull/439)
* ~~[gitlab/revng/orchestra](https://rev.ng/gitlab/revng/orchestra/-/merge_requests/4)~~

mrjackv committed:
ci: introduce ci-regression.sh

Introduce the `ci-regression.sh` script, which will be used for
regression testing and promoting the `develop` branch to `master`.

mrjackv committed:
binary-archives-hook.sh: reduce needed S3 config

Inline parts of the s3cmd config which are not auth-specific or
bucket-specific.

mrjackv committed:
Drop last instances of `#!/bin/bash`

aleclearmind commented on .orchestra/ci/ci-run.sh:
```diff
@@ -106,17 +96,19 @@ fi
+    fi
+    mv "$WHEEL_DIR" "$ORCHESTRA_ROOT/../wheels"
+    break
+done
```
Documentation

mrjackv replied:
As we discussed in the office this will be reverted and the wheels directory will not be created, correct?
In that case this section of code will be dropped and restored back to its original state.

aleclearmind replied:
Yes.

mrjackv replied:
Dropped the code above

aleclearmind commented on .orchestra/ci/ci-regression.sh:
Push `set -euo pipefail`, `SCRIPT_DIR` into `common.sh`.

Also turn `parts/*.sh` into functions into `common.sh`.

We ended up using `source` to perform function calls, which is not very idiomatic.

mrjackv replied:
Implemented

aleclearmind commented on .orchestra/ci/parts/pipeline-create.sh:
```diff
@@ -0,0 +1,77 @@
+#   The target branch to use for the pipeline, defaults to 'master'
+# BINARY_ARCHIVES_HOOKS_FORCE:
+#   If set to '1' the hooks are going to be run regardless if there have been
+#   changes to the binary archives
```
Are we using these anywhere?

Is there some use case in which `ci-hook` sets these?

Let's not have unused features.

mrjackv replied:
* `DOWNSTREAM_PROJECT_TARGET_BRANCH`: I'm ok dropping this
* `BINARY_ARCHIVES_HOOKS_FORCE`: I've added this so that we can trigger hooks without having to do changes like [this](https://github.com/revng/orchestra/commit/51b301d2efcbb508a20e6032f9f01c139289eb2b) every time we want to check if the hooks are working. While "unused" it has a purpose (forcing execution of binary archives hooks) and it's useful (similarly to `llvm_enable_pgo_lto`)

aleclearmind replied:
OK.

mrjackv committed:
.gitlab-ci.yml: sync back from gitlab

Sync the `.gitlab-ci.yml` back from the gitlab CI repo.

mrjackv committed:
mass-testing: introduce check-regression

Add the `check-regression` script, whose job is to read the data from a
previous run of mass-testing and a current one and determine if the
regression passes or not.

aleclearmind commented on .orchestra/ci/binary-archives-hook.sh:
```diff
@@ -7,6 +7,11 @@ set -euo pipefail
+# 1:
+#   File containing the list of changed files in the binary archives, each line
+#   is a different file path
+# COMPONENT_TARGET_BRANCH: target branch the CI has been triggered against
```
Blank line

aleclearmind replied:
Ping

mrjackv replied:
Done

aleclearmind commented on .orchestra/ci/binary-archives-hook.sh:
```diff
@@ -7,6 +7,11 @@ set -euo pipefail
 #
 # Mandatory parameters:
 #
+# 1:
```
```
./binary-archives-hook.sh CHANGED_FILES

CHANGED_FILES
   File containing the list of changed files in the binary archives, each line
   is a different file path

Environment variables:

  COMPONENT_TARGET_BRANCH: target branch the CI has been triggered against

[...]
```

Same elsewhere, the environment variables are list under "parameters", which is confusing.

Also, please document in the header what this script *does*. It only explains when it runs.

aleclearmind replied:
Ping

mrjackv replied:
Done

aleclearmind commented:
I think we need a brief `README.md` summarizing what happens when you push something. I'm losing track of all the things we do.

mrjackv committed:
Delete `.orchestra/ci/get-remote`

Remove the `get-remote` script, as it is unused.

mrjackv committed:
revng mass-testing configure: new input structure

The format of the input file for `revng mass-testing configure` has been
changed. Adapt the `commands.yml` file to work with the changes.

mrjackv committed:
mass-testing: switch to csv

Switch from using a yaml file to a csv file reporting binaries, execution
time and memory usage.

mrjackv commented:
@aleclearmind the changes you requested are implemented but I'm mid-refactoring due to the changes in the regression process that we discussed previously, so this PR will be updated in a while

aleclearmind commented on .orchestra/support/mass-testing/check-regression:
```diff
@@ -0,0 +1,140 @@
+    args = parser.parse_args()
+
+    timeout_margin = os.environ.get("MASS_TESTING_REGRESSION_TIMEOUT_MARGIN", args.timeout_margin)
+    memory_margin = os.environ.get("MASS_TESTING_REGRESSION_MEMORY_MARGIN", args.memory_margin)
```
Arguments take precedence over environment variables, here is the other way around.

mrjackv replied:
As stated [here](https://github.com/revng/revng/pull/439#discussion_r1979683180) these variables need to be set outside of orchestra, e.g.
```
export MASS_TESTING_REGRESSION_TIMEOUT_MARGIN=0.5
orc install mass-testing-regression
```

Having the command-line take precedence over the environment variable would mean that if the value needs to be changed it requires patching the orchestra's `.yml` file, which isn't convenient.

aleclearmind replied:
If you want environment variables to take precedence, do not specify command line arguments.

Having environment variables override command line arguments is not OK.

mrjackv replied:
Done

aleclearmind commented on .orchestra/support/mass-testing/check-regression:
Add a long-ass comment explaining what this script does and the ratio behind it.

mrjackv replied:
Added

mrjackv committed:
Add `ci/mass-testing-weeky`

Add the CI yaml and auxiliary files that allow running mass-testing in
the CI and publish it on the internet.

aleclearmind commented on .orchestra/ci/mass-testing-weekly/aggregate.py:
Add a comment about what this script does

mrjackv replied:
Done

aleclearmind commented on .orchestra/ci/mass-testing-weekly/entrypoint.sh:
```diff
@@ -0,0 +1,88 @@
+#!/bin/bash
```
Fix

mrjackv replied:
Done

aleclearmind commented on .orchestra/ci/mass-testing-weekly/entrypoint.sh:
```diff
@@ -0,0 +1,88 @@
+
+# Install dependencies and the orchestra tool
+sudo apt-get update
+sudo apt-get install -y less nano vim inkscape bc
```
vim? inkscape?

mrjackv replied:
* `inkscape` is needed by `revng mass-testing generate-report` to convert the flamegraphs from svg to png
* `bc` is for the `progress_poller.sh` script
* `vim`, `nano` and `less` are for debugging

aleclearmind commented on .orchestra/ci/mass-testing-weekly/entrypoint.sh:
```diff
@@ -0,0 +1,88 @@
+
+kill -9 "$POLLER_PID"
+
+# If mass-testing failed in any way, sleep forever. This is to allow manual
+# intervention to recover the data (if present), since each run takes multiple
+# days to finish
+if [ "$RC" -ne 0 ]; then
+  while true; do
+    sleep 24h
+  done
+  exit "$RC"
+fi
```
Mmmh, and? We can't SSH into the CI.

mrjackv replied:
Yes you can, it's a bit annoying but it can be done

aleclearmind replied:
1. How?
2. Anyway, infinite wait is a nono. A week tops.

mrjackv replied:
1. I'll write it on mattermost
2. The CI already automatically stops the job if it's over a week, so that's already handled.

aleclearmind commented on .orchestra/ci/mass-testing-weekly/entrypoint.sh:
```diff
@@ -0,0 +1,88 @@
+  exit "$RC"
+fi
+
+# Configure ssh for rsync-ing
```
Clearly state what we're doing here.

We're publishing the mass.rev.ng website, right?

mrjackv replied:
Yes

aleclearmind replied:
OK, add a comment.

mrjackv replied:
Added

aleclearmind commented on .orchestra/ci/mass-testing-weekly/entrypoint.sh:
```diff
@@ -0,0 +1,88 @@
+
+# Run the actual mass-testing
+RC=0
+orc --quiet install mass-testing || RC=$?
```
Who is pushing the results to S3?
Below I think we're pushing to mass.rev.ng.

mrjackv replied:
> Who is pushing the results to S3?

That's taken care by the `orc install mass-testing`, in the sense that the install script pushes to S3

>  Below I think we're pushing to mass.rev.ng.

Exactly

aleclearmind replied:
It seems a bit arbitrary to have the `mass-testing` component to push things to S3 but another script push to mass.rev.ng.

1. Can we move the push to S3 here?
2. Can you summarize what we push to S3?

mrjackv replied:
> It seems a bit arbitrary to have the mass-testing component to push things to S3 but another script push to mass.rev.ng.

`mass-testing` already fiddles with S3 since it has to download the binaries etc. so it made sense to also make it push to S3

1. Sure it can be done
2. Just `regression.autogenerated.csv`, that's the only file that's ever pushed to S3

aleclearmind replied:
What if we put `regression.autogenerated.csv` on mass.rev.ng and avoid having S3 write credentials?

mrjackv replied:
It can be done, but then we rely on `mass.rev.ng` working for regression to work, I'm not sure it's a great dependency
Do note that the credentials used by mass-testing allow to *exclusively* write `regression.autogenerated.csv`. It cannot (over)write any other file

aleclearmind replied:
We really need to unify the infrastructure. We have too many things spread around.

Anyway, let's keep things as they are for now.

aleclearmind commented on .orchestra/ci/mass-testing-weekly/progress_poller.sh:
```diff
@@ -0,0 +1,21 @@
+
+    TOTAL=$(find "$BASE_DIR/inputs" -type f | wc -l)
+    COUNT=$(find "$BASE_DIR/build" -name test-harness.json | wc -l)
+    PERCENT=$(bc <<< "scale=2; ($COUNT * 100) / $TOTAL")
```
Use `$((3 * 3))` or Python.

mrjackv replied:
The issue is that I wanted to truncate to two decimal digits. I guess I'll switch to python

aleclearmind replied:
`python -c` ftw

mrjackv replied:
Done
